src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,109 @@
+    private Ui ui;
+
+    /**
+     * Returns a Duke object that requires the filePath and folderPath to initialize the application,
+     * retrieving tasks stored in the hard disk.
+     *
+     * @param filePath  Relative path to locate the file with the stored tasks.
+     * @param folderPath Relative path to locate the directory storing the file.
+     */
+    public Duke(String filePath, String folderPath) {
+        this.ui = new Ui();
+        this.storage = new Storage(filePath, folderPath);
+        try {
+            this.tasks = new TaskList(storage.load());
+        } catch (IOException e) {
+            this.ui.showLoadingError(e);
+            this.tasks = new TaskList(new ArrayList<String>());
+        }
+    }
+
+    /**
+     * Executes the Duke application, requires user input to perform necessary actions and stops the Duke application
+     * when user inputs the given command to end.
+     * <p></p>
+     * Functions currently supported by Duke: add Task, remove Task, mark Task, unmark Task, delete Task, list, bye.
+     * @throws IOException On input error.
+     * @see IOException
+     */
+    public void run() throws IOException {
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            ArrayList<String> commandInfoList;
+            try {
+                commandInfoList = Parser.parse(ui.readCommand());
+            } catch (IncorrectNoOfArgumentException e) {
+                System.out.println(e);
+                continue;
+            }
+            if (commandInfoList.size() != 0) {
+                switch (commandInfoList.get(0)) {
+                case "bye":
+                    isExit = true;
+                    ui.showFarewellMessage();
+                    break;
+                case "list":
+                    ui.printCommand(tasks.list());
+                    break;
+                case "find":
+                    ui.printCommand(tasks.find(commandInfoList.get(1)));
+                    break;
+                case "mark":
+                    ui.printCommand(tasks.markTask(commandInfoList, this.storage));
+                    break;
+                case "unmark":
+                    ui.printCommand(tasks.unmarkTask(commandInfoList, this.storage));
+                    break;
+                case "delete":
+                    ui.printCommand(tasks.deleteTask(commandInfoList, this.storage));
+                    break;
+                case "error":
+                    break;
+                default:
+                    try {
+                        ui.printCommand(tasks.addTask(commandInfoList, this.storage));
+                    } catch (DateTimeParseException e) {
+                        System.out.println("Invalid inputs!\n");
+                        System.out.println("Please enter your date & time in the format: YYYY-MM-DD HH:MM \n");
+                        System.out.println("Please also ensure they are valid values!\n");
```

Comment:
This is a very long and deeply nested function. Maybe you could consider abstracting the commands into different classes?

Reply:
I agree with you on the creation of another class possibly command class with subclasses dealing with the respective command. This would probably be an area I would work on and release for future commits!